### PR TITLE
FNV support

### PR DESCRIPTION
**Pull request by mklokocka**
Add things I believe to be required for full FNV support.

Namely:

1) A default profile for FNV including patterns for selecting what to pack, optimize, etc.
2) Disabled mesh optimizations for FNV.
3) Ability to specify what to pack using patterns.
4) Option to create `.override` files for each archive created.

### CODE REVIEW

File: src/bsa_process.cpp [hunk 1]
```diff
@@ -68,4 +71,20 @@ void write_single_archive(const btu::Path &directory_path,
...
+        const auto filename      = archive_path.value().stem().u8string() + u8".override";
+        const auto override_path = parent_path / filename;
+
+        if (exists(override_path))
+        {
+            PLOGI << override_path.string() << " already exists";
+        }
+        else
+        {
+            std::ofstream out(override_path);
```
Comment (Guekka) [inline src/bsa_process.cpp:87]: Use `btu::common::write_file`

File: src/bsa_process.cpp [hunk 2]
```diff
@@ -68,4 +71,20 @@ void write_single_archive(const btu::Path &directory_path,
         const auto count = remove_files_from_directory(directory_path, relative_paths);
         PLOGI << std::format("Attempted to remove {} files, {} were removed", relative_paths.size(), count);
     }
+
+    if (make_override)
+    {
+        const auto parent_path   = archive_path.value().parent_path();
+        const auto filename      = archive_path.value().stem().u8string() + u8".override";
+        const auto override_path = parent_path / filename;
```
Comment (Guekka) [inline src/bsa_process.cpp:79]:
```suggestion
        const auto override_path = std::move(archive_path).replace_extension(".override");
```

File: src/gui/AdvancedBSAModule.cpp [hunk 3]
```diff
@@ -46,6 +46,7 @@ void AdvancedBSAModule::settings_to_ui(const Settings &settings)
     ui_->dontMakeLoaded->setChecked(!profile.bsa_make_dummy_plugins);
     ui_->dontRemoveFiles->setChecked(!profile.bsa_remove_files);
     ui_->dontCompress->setChecked(!profile.bsa_allow_compression);
+    ui_->makeOverrides->setChecked(profile.bsa_make_overrides);
```
Comment (Guekka) [inline src/gui/AdvancedBSAModule.cpp]: We should only show this button when the game supports override, don't you think?

File: src/settings/profile.hpp [hunk 4]
```diff
@@ -101,7 +101,41 @@ class Profile
...
+            profile.bsa_make_overrides    = true;
+
+            profile.base_per_file_settings_.pack = false;
+
+            auto fnv_lod_textures         = PerFileSettings::make_base(game);
+            fnv_lod_textures.pattern      = Pattern(u8".*lod(_[nsdgpe])?\\.dds", Pattern::Type::Regex);
+            fnv_lod_textures.tex_optimize = OptimizeType::None;
+
+            auto fnv_ui_textures         = PerFileSettings::make_base(game);
+            fnv_ui_textures.pattern      = Pattern(u8"textures/interface/.*\\.dds", Pattern::Type::Regex);
```
Comment (Guekka) [inline src/settings/profile.hpp:114]: Use a wildcard:
```suggestion
            fnv_ui_textures.pattern      = Pattern(u8"textures/interface/*.dds", Pattern::Type::Wildcard);
```

File: src/settings/profile.hpp [hunk 5]
```diff
@@ -101,7 +101,41 @@ class Profile
...
+            fnv_ui_textures.tex_optimize = OptimizeType::None;
+
+            auto fnv_dds         = PerFileSettings::make_base(game);
+            fnv_dds.pattern      = Pattern(u8".*\\.dds", Pattern::Type::Regex);
```
Comment (Guekka) [inline src/settings/profile.hpp:118]:
```suggestion
            fnv_dds.pattern      = Pattern(u8"*.dds", Pattern::Type::Wildcard);
```

File: src/settings/profile.hpp [hunk 6]
```diff
@@ -101,7 +101,41 @@ class Profile
...
+            fnv_dds.tex.compress = true;
```
Comment (Guekka) [inline src/settings/profile.hpp:119]: Why? FNV only supports older compression formats that create visual artifacts. Compression should probably not be the default

File: src/settings/profile.hpp [hunk 7]
```diff
@@ -101,7 +101,41 @@ class Profile
...
+            fnv_dds.tex.resize   = btu::tex::Dimension{2048, 2048};
```
Comment (Guekka) [inline src/settings/profile.hpp:120]: Why? Is there a reason to resize by default?

File: src/settings/profile.hpp [hunk 8]
```diff
@@ -101,7 +101,41 @@ class Profile
...
+
+            auto fnv_tga         = PerFileSettings::make_base(game);
+            fnv_tga.pattern      = Pattern(u8".*\\.tga", Pattern::Type::Regex);
```
Comment (Guekka) [inline src/settings/profile.hpp:123]:
```suggestion
            fnv_tga.pattern      = Pattern(u8"*.tga", Pattern::Type::Wildcard);
```

File: src/settings/profile.hpp [hunk 9]
```diff
@@ -101,7 +101,41 @@ class Profile
...
+            fnv_tga.tex_optimize = OptimizeType::None;
+
+            auto fnv_meshes         = PerFileSettings::make_base(game);
+            fnv_meshes.pattern      = Pattern(u8".*\\.nif", Pattern::Type::Regex);
```
Comment (Guekka) [inline src/settings/profile.hpp:127]:
```suggestion
            fnv_meshes.pattern      = Pattern(u8"*.nif", Pattern::Type::Wildcard);
```

File: src/settings/profile.hpp [hunk 10]
```diff
@@ -101,7 +101,41 @@ class Profile
...
+            fnv_meshes.nif_optimize = OptimizeType::None;
+
+            auto fnv_sounds    = PerFileSettings::make_base(game);
+            fnv_sounds.pattern = Pattern(u8".*\\.(ogg|wav)", Pattern::Type::Regex);
```
Comment (Guekka) [inline src/settings/profile.hpp:131]:
```suggestion
            fnv_sounds.pattern = Pattern(u8"*.[ow][ga][gv]", Pattern::Type::Wildcard);
```

File: src/settings/profile.hpp [hunk 11]
```diff
@@ -101,7 +101,41 @@ class Profile
 
         if (game == btu::Game::FNV)
```
Comment (Guekka) [inline src/settings/profile.hpp:102]: Maybe this should be moved to a separate function

**Review by Guekka — COMMENTED** [review]
First pass review, I haven't looked at the GUI changes yet
It looks good. Please prefer wildcards over regex when possible, they have better performance

File: src/gui/AdvancedBSAModule.cpp (hunk 3 above)
Comment (mklokocka) [reply src/gui/AdvancedBSAModule.cpp]: Yep, that definitely makes sense.

File: src/settings/profile.hpp (hunk 6 above)
Comment (mklokocka) [reply src/settings/profile.hpp:119]: When approaching the default profile I tried to set it up with what I perceive to be the “golden standard” for FNV mods in the community. Which as of now is to use compressed textures at most 2k in resolution. 

Many retexture mods are hilariously oversized (4k+), sometimes even uncompressed. Interestingly nearly all textures >2k are mipmapped by the engine to <2k so using them has no benefit at all.

File: src/settings/profile.hpp (hunk 7 above)
Comment (mklokocka) [reply src/settings/profile.hpp:120]: See above reply, basically anything over 2k doesn’t make sense for the game.

File: src/settings/profile.hpp (hunk 11 above)
Comment (mklokocka) [reply src/settings/profile.hpp:102]: Agreed, I just didn’t figure out a place where it would make the most amount of sense.

**Comment by mklokocka** [comment]
Merged your TODO commit into others (and split out the dependency updates). Notably I changed the interface texture pattern back to regex to be able to catch subdirectories.

Also hidden the override toggle for non-FNV games.

**Comment by Guekka** [comment]
> I changed the interface texture pattern back to regex to be able to catch subdirectories

Wildcards can match subdirectories. Maybe my explanation of them was not clear?

**Review by Guekka — APPROVED** [review]
Looks good. I'm just not sure I like the split of the BSA module, but it makes sense

**Comment by mklokocka** [comment]
Re wildcards: How? There really ought to be some description for how wildcard patterns work. 😅 I thought they were like unix blobs where `*` only catches stuff between directory separators.

Re BSA module split: I tend to agree it's a bit weird having it split, on the other hand it makes semantic sense, since the "pack" setting is tied to patterns more than being a "general" thing.

**Comment by Guekka** [comment]
> Re wildcards: How? There really ought to be some description for how wildcard patterns work. 😅 I thought they were like unix blobs where * only catches stuff between directory separators.

The algorithm is [here](https://github.com/Guekka/bethutil/blob/de993e54a0522ddb4788c4951aec04ca45255e2d/include/btu/common/string.hpp#L284)
It's more like a subset of wildcards, which is, I agree, confusing. So `*` matches anything

**Comment by mklokocka** [comment]
> It's more like a subset of wildcards, which is, I agree, confusing. So * matches anything
Just tested it again, and I know what the "problem" is - wildcard patterns do not ignore case, unlike regex patterns. So using your proposed wildcard pattern `textures/interface/*.dds` - `textures/interface/something/tex.dds` is correctly skipped, while `textures/Interface/something/tex.dds` is processed.

**Comment by Guekka** [comment]
> wildcard patterns do not ignore case
Thanks for trying. Indeed, the case insensitive flag is missing [here](https://github.com/Guekka/cathedral-assets-optimizer/blob/dev/src/settings/per_file_settings.hpp#L55)

In the meantime, we can keep your regex. Tell me when I can merge

**Comment by mklokocka** [comment]
As long as you are happy with the changes we can merge. 👍

**Pull request merged**